[DOCS] Add clarifications to mapping type (#7100)
ethereum/solidity: docs/types/mapping-types.rst

Mapping Types
=============

> You declare mapping types with the syntax ``mapping(_KeyType => _ValueType) _VariableModifiers _VariableName``.

Reviewer: This is how you declare a variable of mapping type, not how you declare a mapping type.

Maybe `declare` is also the wrong word, `use` may be better.

Mapping Types
=============

Mapping types use the syntax ``mapping(_KeyType => _ValueType)`` and variables

> are declared as a mapping type using the syntax ``mapping (_KeyType => _ValueType) _VariableModifiers _VariableName``.

Reviewer: It's probably not too useful to keep this.

Reviewer: I still do not think it is useful to keep this.

Author: Which bit? It was here already, and from the conversations I've had around, people are really confused by the mapping syntax, so I'd love to try and be as explicit and clear as possible. Any suggestions to keep that clarity but express it differently?

Reviewer: I think it would be better to give an example how to declare a specific variable instead of just adding random placeholders whose meaning is not explained.

of contract functions that are publicly visible.

You can mark state variables of mapping type as ``public`` and Solidity creates a

> :ref:`getter <visibility-and-getters>` for you. The ``_KeyType`` is the parameter for the getter.

Reviewer: Please keep "becomes a", because it may not be the only parameter.

of contract functions that are publicly visible.

You can mark state variables of mapping type as ``public`` and Solidity creates a
:ref:`getter <visibility-and-getters>` for you. The ``_KeyType`` becomes a parameter for the getter.
If ``_ValueType`` is a value type or a struct, the getter returns ``_ValueType``.
If ``_ValueType`` is an array or a mapping, the getter has one parameter for
each ``_KeyType``, recursively.

> In the example below, the ``MappingExample`` contract defines a public ``balances``

Reviewer: Please also add a mapping of a mapping to the example.

> mapping, with the key type an ``address``, and a value type a ``uint``, mapping

Reviewer: Isn't it mapping the address to the uint?

Author: I think that is what I was trying to say, but I'll make it clearer.

of contract functions that are publicly visible.

You can mark state variables of mapping type as ``public`` and Solidity creates a
:ref:`getter <visibility-and-getters>` for you. The ``_KeyType`` becomes a parameter for the getter.
If ``_ValueType`` is a value type or a struct, the getter returns ``_ValueType``.
If ``_ValueType`` is an array or a mapping, the getter has one parameter for
each ``_KeyType``, recursively.

In the example below, the ``MappingExample`` contract defines a public ``balances``
mapping, with the key type an ``address``, and a value type a ``uint``, mapping
an Ethereum address to an unsigned integer value. As ``uint`` is a value type, the getter

> returns a value that matches the type, which you can see in the ``MappingUser``

Reviewer: "of that type" (how does a value match a type?)

parameter for the getter. If ``_ValueType`` is a value type or a struct,
the getter returns ``_ValueType``.

> If ``_ValueType`` is an array or a mapping, the getter has one parameter for

Reviewer: This doesn't really make sense. Which `_KeyType`? The getter has additional parameters as needed to access the final element in the mapping / array structure.

Author: @chriseth it refers to the paragraph a couple above that uses the same analogies, but I'll make the connection between the two paragraphs clearer.

Reviewer 2: well, the keytype is mentioned in line 7. But for clarification we could write "the mappings keytype" / "the keytype of the mapping"
On that note, maybe a cod example is useful here, too

Author: There is a code example a paragraph later, might be better to view the overall changes to see everything in context and let me know if you think anything is still needed.

Reviewer: Ah sorry, now I see it!

each ``_KeyType``, recursively.

In the example below, the ``MappingExample`` contract defines a public ``balances``
mapping, with the key type an ``address``, and a value type a ``uint``, mapping

> an integer value to an Ethereum address. As ``uint`` is a value type, the getter

Reviewer: Mapping an address to an unsigned integer.

        }
    }

The example below uses a mapping type inside another mapping type. The important

> point to remember are that the mapping inside the mapping is not the return value,

Reviewer 3: That should be ``The important point to remember is`` not ``are``... but also: "the return value" of what? I'm not sure I understand the purpose of this sentence :-).

Author: Corrected that. The point is to clarify that the returned value isn't the mapping inside the mapping, but the value inside the nested mapping. That's what I understood from you anyway?

Reviewer 3: When you're talking about "the returned value", you have to say what you are talking about that returns something - the index access of the outer mapping? The index access of the outer mapping in fact *does* return "the mapping inside the mapping" - only the second index access yields the value contained in the inner mapping...

```
contract C {
    mapping(address => mapping(address => bool)) public allowances;
    function f(address X, address Y) public view {
        mapping(address => bool) storage allowancesForX = allowances[X]; // first index access is a mapping
        bool allowanceForXandY = allowances[X][Y]; // two index accesses give the contained value
        bool allowanceForXandY2 = allowancesForX[Y]; // which is the same as accessing the contained mapping at Y
        assert(allowanceForXandY == allowanceForXandY2);
    }
}

```

> but the variable inside it is.

Reviewer: "value" instead of 'variable' for both?

> ::

Author: @chriseth Myself and @ekpyron had a chat about an example for a mapping with a mapping and struggled to think of many great examples. What do you think of this one? Then I'll polish the text.

Reviewer: looks good!

    pragma solidity >=0.4.0 <0.7.0;

>     contract MappingExample {

Reviewer: Do you think a full credit / allowance example would be too long?

Author: @chriseth Possibly, but how much do you think the full example relates to mappings and how they work? Or maybe a reduced example of what you suggest that highlights those aspects? I assume based on the Zeppelin link from above?

Reviewer: I think you can use the following functions and also slim them down a little:
 - approve
 - transferFrom 
 - allowances

>         mapping(address => mapping(address => bool)) public validBalances;

Reviewer: Doesn't erc20 use something like `allowances`?

Author: It does, but I'm not sure what you mean? This isn't really an ERC20 implementation, and the data structure here is different too.

Reviewer: No, it's exactly the same - and I think it would be a good example: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L36

        function transfer(address sender, address recipient) view public {
            require(validBalances[sender][recipient]);
        }

        function allowTransfers(address recipient) public {
            validBalances[msg.sender][recipient] = true;
        }

    }

    contract MappingUser {
        function f() public returns (bool) {
            MappingExample m = new MappingExample();

>             m.transfer(address(this),address(this));

Reviewer: ```suggestion
            m.transfer(address(this), address(this));
```

Reviewer: maybe at least transfer from this to msg.sender? Or provide an address as a parameter of `f()`? Then you do not need the type conversions.

>             return m.validBalances(address(this),address(this));

Reviewer: ```suggestion
            return m.validBalances(address(this), address(this));
```

        }
    }

The example below is a simplified version of an `ERC20 token <https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol>`_.
``_allowances`` is an example of a mapping type inside another mapping type.

> The example below uses ``_allowances`` to record the amount of a transaction and the two addresses it was between.

Reviewer: ```suggestion
The example below uses ``_allowances`` to record the amount someone else is allowed to withdraw from your account.
```

::

    pragma solidity >=0.4.0 <0.7.0;

    contract MappingExample {

        mapping (address => uint256) private _balances;
        mapping (address => mapping (address => uint256)) private _allowances;

        event Transfer(address indexed from, address indexed to, uint256 value);
        event Approval(address indexed owner, address indexed spender, uint256 value);

        function allowance(address owner, address spender) public view returns (uint256) {
            return _allowances[owner][spender];
        }

        function transferFrom(address sender, address recipient, uint256 amount) public returns (bool) {
            _transfer(sender, recipient, amount);

>             _approve(sender, msg.sender, amount);

Reviewer: Why do you call approve here? It should rather check that the allowance is large enough.

Author: @chriseth This came from the Zeppelin code…

Reviewer: But it does something else now.

        event Transfer(address indexed from, address indexed to, uint256 value);
        event Approval(address indexed owner, address indexed spender, uint256 value);

        function allowance(address owner, address spender) public view returns (uint256) {
            return _allowances[owner][spender];
        }

        function transferFrom(address sender, address recipient, uint256 amount) public returns (bool) {
            _transfer(sender, recipient, amount);
            _approve(sender, msg.sender, amount);
            return true;
        }

        function approve(address spender, uint256 amount) public returns (bool) {
            _approve(msg.sender, spender, amount);
            return true;
        }

        function _transfer(address sender, address recipient, uint256 amount) internal {
            require(sender != address(0), "ERC20: transfer from the zero address");
            require(recipient != address(0), "ERC20: transfer to the zero address");

            _balances[sender] -= amount;
            _balances[recipient] += amount;
            emit Transfer(sender, recipient, amount);
        }

>         function _approve(address owner, address spender, uint256 amount) internal {

Reviewer: I think it is better to combine this function with `approve` and handle the approval reduction in transferFrom manually.

Author: I think this is what you're asking for…
